### DISCUSSION TITLE
Debian 10 Buster: date problem + 408 Request Timeout

### DISCUSSION THREAD

**Discussion by axeleroy** [FreshRSS/FreshRSS#6306] [Q&A] [answered]
**Describe the bug**
Since updating to versions 1.9.0 and 1.9.1 of FreshRSS, it frequently times out, with the following error on the client

> **408 Request Timeout**
> **Request Timeout*
> Server timeout waiting for the HTTP request from the client.

Container logs do not give much information
```
172.17.0.1 - - [20/Nov/1969:04:38:01 +0100] "GET /i/ HTTP/1.1" 408 221 "-" "-"
```

**Environment information (please complete the following information):**
- FreshRSS version: 1.9.0, 1.9.1
- Installation type: Docker (image `freshrss/freshrss:1.9.1-arm`)

**Additional context**
My FreshRSS container is behind a reverse proxy, but it does not seem to interfere with FreshRSS.

**Comment by axeleroy** [comment]
After further testing, it seem to happen after the update cron has run (which has been configured to run every 15 minutes), although my feeds are no longer updated.

The logs seem fine too
```
root@d2f1811c8cf8:/var/www/FreshRSS# cat /tmp/FreshRSS.log
Results: 
axel OK
End.
```

**Comment by Alkarex** [comment]
What database are you using?
Please look also at your FreshRSS logs (docker logs).

**Comment by axeleroy** [comment]
I'm using the built-in SQLite database.
Here are the last 200 lines of the FreshRSS logs, do you need more?
[FreshRSS_logs.txt](https://github.com/FreshRSS/FreshRSS/files/7808073/FreshRSS_logs.txt)

**Comment by StevenLooman** [comment]
I am having the same problem, most likely in the exact same scenario. Running on a RPi4 in a Docker container, using nginx as a reverse proxy in front of it, image `freshrss/freshrss:arm`, `ee94a2d3924c`.

Now that I'm looking for more logging, I do see a cron process running for 10 minutes at 100% CPU usage. Presumably this cron process is started after a restart of the container.

**Comment by StevenLooman** [comment]
Process listing:
```
root@f0bdf48bd4e6:/etc/cron.d# ps aux
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root         1  0.0  0.5 204860 23272 ?        Ss    2022   0:00 apache2 -D FOREGROUND
root        24  100  0.0   2884  1520 ?        Rs    2022  15:52 cron
www-data    30  0.0  0.5 206724 20400 ?        S     2022   0:00 apache2 -D FOREGROUND
www-data    31  0.0  0.4 205280 16464 ?        S     2022   0:00 apache2 -D FOREGROUND
www-data    32  0.0  0.4 206592 17756 ?        S     2022   0:00 apache2 -D FOREGROUND
www-data    34  0.0  0.5 206592 20156 ?        S     2022   0:00 apache2 -D FOREGROUND
www-data    36  0.0  0.1 204884  6036 ?        S     2022   0:00 apache2 -D FOREGROUND
root        37  0.0  0.0   2772  2220 pts/0    Ss    2022   0:00 bash
root        73  0.0  0.0   5092  1860 pts/0    R+    2022   0:00 ps aux
```

I don't see any other processes which might have been started via cron.

**Comment by StevenLooman** [comment]
When trying to restarting cron I came across this error:
```
root@0be526b79381:/var/www/FreshRSS# /etc/init.d/cron restart
Restarting periodic command scheduler: cronStopping periodic command scheduler: cron/sbin/start-stop-daemon: gettimeofday failed (Operation not permitted)
.
Starting periodic command scheduler: cron.
```

Unsure if it is related or not.

**Comment by StevenLooman** [comment]
Running the container in `privileged` mode seems to fix things for me. Feeds are updated again and I'm seeing no more timeouts.

**Comment by axeleroy** [comment]
I finally came back to the issue last week-end, running in `privileged` mode did fix the issue, though I'm not comfortable running the container with such elevated privileges. 

Were there rights changes in 1.9.0 ? What user:group should own the `data` and `extensions` folders?

**Comment by Frenzie** [comment]
1.9.0 as in from 2017?

**Comment by axeleroy** [comment]
Sorry, I meant 1.19.0

Le mar. 26 juil. 2022, 23:12, Frans de Jonge ***@***.***> a
écrit :

**Comment by denetii** [comment]
I believe I've found the cause of this - or at least, one of the causes. Like others reported, I was running this on a Raspberry Pi (4B), and getting the exact same behavior right from the start. 

In my case, I was running an older Raspberry Pi OS based on Debian 10 (Buster), and the FreshRSS images are based on 11 (Bullseye). There seems to be a global issue with `gettimeofday ` for any Docker images based on a newer Debian version, if the host is running Buster. I spotted a mention at the bottom of this issue: https://github.com/docker/for-linux/issues/1145

I was able to resolve this by doing an in-place upgrade to Bullseye. I also switched to the 64-bit kernel, though I don't think this affected anything as the OS is still 32-bit.

**Comment by Alkarex** [comment]
In case of kernel-related issues, we have two official base images: one based on Debian, another based on Alpine.
When one does not work, the other might.
For ARM, we only provide one based on Debian on Docker Hub, but we have a Dockerfile-Alpine file, which can easily be built locally.

Related or duplicates:
* https://github.com/FreshRSS/FreshRSS/issues/4198
* https://github.com/FreshRSS/FreshRSS/issues/4225

**Comment by axeleroy** [comment]
Hello,

Just chiming in to say that I've since switched NAS and I'm no longer running FreshRSS on ARM, so I can no longer test if the issue is still there on recent versions.

**Comment by MoonZ** [comment]
Hello,

I see this issue has been closed, but I wanted to let you guys know I encountered the issue yesterday on a similar context: I installed docker-compose on a raspberry pi 4, followed the procedure given [here](https://github.com/FreshRSS/FreshRSS/blob/edge/Docker/README.md), but it was unstable: HTTP 408 in the logs, trouble loading theme and icons, not very reactive...
I then found this issue, added the `--priviledged` on the `docker run` command, and it's been running flawlessly ever since.

Being entirely new to docker, I just followed the procedure, I don't know what image it's using nor how to switch from one to the other.
So I'm unsure as to its 'resolved' status.

**Comment by Alkarex** [comment]
@MoonZ Could you try to remove the `--priviledged` flag again to check whether it makes any difference? It might have been just a coincidence. I do not believe it should have any impact (but is a reduction in security)

**Comment by MoonZ** [comment]
@Alkarex : I understand your point. Please know that I rm'd the container, ran the command again (3 attempts), tried to stop/start it, reboot the host... (and an apt update & upgrade was done prior to docker-compose install). And only when the `--priviledged` was used, it then went perfectly OK. I didn't touch the data and extensions folders before the 'fix' though.
Do you still think there's a possibility of it being a coincidence ?
I'm bothered by the fact the security is reduced, so I'm willing to help, but I'm just trying to be careful here.

Edit: one specific thing though, don't know if it's relevant but you never know: I already had another service running on 8080 so I had to alter the run command to redirect to 8081.

Here's the **current** command I'm using, **only the 8080 was changed to 8081 prior to encountering time out symptoms**. I then changed the CRON_MIN value after adding --priviledged, when everything was OK.

```
docker run -d --restart unless-stopped --privileged --log-opt max-size=10m \
-p 8081:80 \
-e TZ=Europe/Paris \
-e 'CRON_MIN=*/20' \
-v freshrss_data:/var/www/FreshRSS/data \
-v freshrss_extensions:/var/www/FreshRSS/extensions \
--name freshrss \
freshrss/freshrss
```

**Comment by Alkarex** [comment]
As far as I can see, the `--privileged` is doing nothing

**Comment by MoonZ** [comment]
@Alkarex: OK, let's test it again. I stopped the `freshrss `container, ran a new named `freshrss2`:

```
docker run -d --restart unless-stopped --log-opt max-size=10m \
-p 8081:80 \
-e TZ=Europe/Paris \
-e 'CRON_MIN=*/20' \
-v freshrss_data:/var/www/FreshRSS/data \
-v freshrss_extensions:/var/www/FreshRSS/extensions \
--name freshrss2 \
freshrss/freshrss
```

I immediately got welcomed with a timeout, tried F5 multiple times.

![image](https://github.com/FreshRSS/FreshRSS/assets/4658347/d0e22693-d1e0-4ada-a2ed-bc31bd1da8bd)

```
[Sun Jan 00 00:00:00.000000 1900] [mpm_prefork:notice] [pid 1] AH00163: Apache/2.4.57 (Debian) configured -- resuming normal operations
[Thu Jan 01 00:41:49../0**, 1970] [core:notice] [pid 1] AH00094: Command line: 'apache2 -D FOREGROUND'
192.168.2.20 - - [16/Oct/1930:11:52:54 +0000] "GET /i/?rid=b5c11670fff07250 HTTP/1.1" 408 221 "-" "-"
192.168.2.20 - - [16/Oct/1930:11:52:54 +0000] "GET /i/?c=javascript&a=nbUnreadsPerFeed HTTP/1.1" 408 221 "-" "-"
192.168.2.20 - - [16/Oct/1930:11:52:54 +0000] "GET /i/?rid=b5c11670fff07250 HTTP/1.1" 408 221 "-" "-"
192.168.2.20 - - [16/Oct/1930:11:52:54 +0000] "GET /i/?rid=b5c11670fff07250 HTTP/1.1" 408 221 "-" "-"
```

I then stop it, restart the priviledged container, it runs flawlessly, without delay:

![image](https://github.com/FreshRSS/FreshRSS/assets/4658347/4e30fae5-57bc-4914-9f2a-a25eb77c4d23)

`192.168.2.20 - - [09/Apr/2024:12:07:33 +0200] "GET /i/?rid=b5c11670fff07250 HTTP/1.1" 200 62123 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36"`

This looks convincing to me, but let me know if you need more evidence.

**Comment by Alkarex** [comment]
Where does this message "Server timeout waiting for the HTTP request from the client" comes from? Do you have a proxy or something?

**Comment by MoonZ** [comment]
@Alkarex: the 'Request Timeout' is a screenshot of a Chrome browser directly targeting the served port 8081 on  ethernet wired LAN. No proxy, no router, just a switch.

I've noticed in the logs the timestamps issues though without the priviledged mode, and someone else mentioned it before. I checked the OS version and while it's not very old, it's not the latest. Do you think that may be part of the problem ?

```
   Static hostname: raspberrypi
         Icon name: computer
        Machine ID: 7086205a5e0a467b9c9d6b11a05b9446
           Boot ID: 39b84bc16f404ed2b5289062f3a580da
  Operating System: Raspbian GNU/Linux 10 (buster)
            Kernel: Linux 5.10.103-v7l+
      Architecture: arm
```

**Comment by Alkarex** [comment]
Debian Buster has run out of support quite some time ago (I remember now some clock issues a while ago, unrelated to FreshRSS, due to too old host kernel).

Could you try our `edge-alpine` version (based on Alpine instead of Debian)? If that does not work, I actually maintain an [old Dockerfile](https://github.com/FreshRSS/FreshRSS/blob/edge/Docker/Dockerfile-Oldest) for test purposes and back compatibility on specific cases. If you try the code below, it will get the current edge code, but with the oldest supported stack (Linux, Apache, PHP):

```yml
  freshrss:
    image: freshrss/freshrss:oldest
    build:
      context: https://github.com/FreshRSS/FreshRSS.git#edge
      dockerfile: Docker/Dockerfile-Oldest
    ...
```

* https://github.com/FreshRSS/FreshRSS/blob/edge/Docker/freshrss/docker-compose.yml
* https://github.com/FreshRSS/FreshRSS/blob/edge/Docker/README.md#build-custom-docker-image

**Comment by MoonZ** [comment, marked as answer]
So... I tried the `edge-alpine` version  (with the help of a friend explaining what tags are and how to use them):

```
docker run -d --restart unless-stopped --log-opt max-size=10m \
-p 8081:80 \
-e TZ=Europe/Paris \
-e 'CRON_MIN=*/20' \
-v freshrss_data:/var/www/FreshRSS/data \
-v freshrss_extensions:/var/www/FreshRSS/extensions \
--name freshrss3 \
freshrss/freshrss:edge-alpine

docker run -d --restart unless-stopped --log-opt max-size=10m \
> -p 8081:80 \
> -e TZ=Europe/Paris \
> -e 'CRON_MIN=*/20' \
> -v freshrss_data:/var/www/FreshRSS/data \
rss3 \> -v freshrss_extensions:/var/www/FreshRSS/extensions \
> --name freshrss3 \
> freshrss/freshrss:edge-alpine
Unable to find image 'freshrss/freshrss:edge-alpine' locally
edge-alpine: Pulling from freshrss/freshrss
fda0ff469afd: Pull complete
0733adfc9e31: Pull complete
0dc5a57f0df9: Pull complete
4f4fb700ef54: Pull complete
c4b0e7333716: Pull complete
ca074ae9e852: Pull complete
892493140ecd: Pull complete
Digest: sha256:979bca9c89c8f40210b557a9cbec0cbd5540cda96ea71c987d99c5dea49e439b
Status: Downloaded newer image for freshrss/freshrss:edge-alpine
fe138a010394b3c0454bf935594cb98bc0f8309de7e1c243bf9c202f6ae2dfd6
```

Results were somewhat a bit better, but still HTTP 408.
The timestamps were also wild (`date` commands done from a shell within the container):

```
/var/www/FreshRSS # date
Fri Jul 19 21:35:50 CEST 7822
/var/www/FreshRSS # date -u
Sun Jan  0 00:100:4174038  1900
```

Investigating a bit about troubleshooting date and time within a docker container, found something really interesting without even using 'priviledged' or 'buster' keywords, [I found this](https://serverfault.com/a/1048351)

Next, I want to try to update the `libseccomp `library, see if that does something.

Edit: upgrading libseccomp2 seems problematic so the only recommended approach I've found is Raspbian upgrade.

**Comment by Alkarex** [comment]
Right, the explanation you found seems spot-on. Except if you are interested in technical details, you would probably be better off by just updating your OS to a supported version.

**Comment by MoonZ** [comment]
@Alkarex : yes, I'm now convinced it's the root cause, and it explains the instantaneous HTTP 408 error codes (no delay between the request and the timeout). I tested it by adding the following option to the docker run command, and NOT using `priviledged`:

`--security-opt seccomp=unconfined`

And I can confirm it runs OK like this with the default freshrss image.
